src/shamir.rs

```diff
@@ -0,0 +1,419 @@
+        assert!(matches!(
+            SecretSharing::reconstruct(&shares[0..1], &LagrangePolynomial::new(n).unwrap()),
+            Err(Error::BadPolynomial {
+                polynomial_degree: 3,
+                points_count: 1
+            })
+        ));
+    }
+
+    #[test]
+    fn fails_if_not_enough_coefficients() {
+        let k = NonZeroU8::new(2).unwrap();
+        let n = NonZeroU8::new(3).unwrap();
+        let shamir = SecretSharing::new(k, n).unwrap();
+
+        let shares = shamir.split(Fp31::from(42_u128), thread_rng());
+        assert!(matches!(
+            SecretSharing::reconstruct(&shares, &LagrangePolynomial::new(k).unwrap()),
+            Err(Error::BadPolynomial {
+                polynomial_degree: 2,
+                points_count: 3
+            })
+        ));
+    }
+
+    #[test]
+    fn can_reject_bad_k_and_n() {
+        for k in [1_u8, 2, 3] {
+            let n = NonZeroU8::new(max(1, k - 1)).unwrap();
+            let k = NonZeroU8::new(k).unwrap();
+            let r = SecretSharing::new(k, n);
+            assert!(matches!(r, Err(Error::BadSharingScheme { .. })));
+        }
+    }
+
+    #[test]
+    fn fails_to_reconstruct_secret_with_duplicated_shares() {
+        let (k, n) = (NonZeroU8::new(2).unwrap(), NonZeroU8::new(3).unwrap());
+        let rng = StepRng::new(1, 1);
+        let secret = Fp31::from(5_u128);
+
+        let sharing = SecretSharing::new(k, n).unwrap();
+
+        let mut shares = sharing.split(secret, rng);
+        shares[2] = shares[1].clone();
+
+        assert_ne!(
+            secret,
+            SecretSharing::reconstruct(&shares, &LagrangePolynomial::new(n).unwrap()).unwrap()
+        );
+    }
+
+    #[test]
+    fn fails_to_reconstruct_secret_with_forged_shares() {
+        let (k, n) = (NonZeroU8::new(2).unwrap(), NonZeroU8::new(3).unwrap());
+        let rng = StepRng::new(1, 1);
+        let secret = Fp31::from(5_u128);
+
+        let sharing = SecretSharing::new(k, n).unwrap();
+
+        let mut shares = sharing.split(secret, rng);
+        shares[2] = Share { y: shares[1].y };
+
+        assert_ne!(
+            secret,
+            SecretSharing::reconstruct(&shares, &LagrangePolynomial::new(n).unwrap()).unwrap()
+        );
+    }
+
+    #[test]
+    fn rejects_polynomials_larger_than_field_modulus() {
+        let bad_n = NonZeroU8::new(Fp31::PRIME).unwrap();
+        assert!(matches!(
+            LagrangePolynomial::<Fp31>::new(bad_n),
+            Err(Error::FieldSizeError {
+                polynomial_degree: _,
+                field_size: _
+            })
+        ));
+    }
+
+    //
+    // Randomized tests
+    //
+
+    #[derive(Debug)]
+    struct ShareReconstructInput {
+        rng_seed: u64,
+        k: u8,
+        n: u8,
+    }
+
+    impl ShareReconstructInput {
+        fn gen() -> impl Strategy<Value = Self> {
+            (2u8..Fp31::PRIME)
+                .prop_flat_map(|v| {
+                    let k = Just(v);
```

Comment:
What does "Just(v)" do?

Reply:
it is basically a proptest strategy that always produces a single value (given as input). The reason why it is used here is because I wanted to enforce the invariant `n >= k` otherwise input does not make much sense and will be rejected by the secret sharing